src/Features/Core/Portable/Completion/CompletionItem.cs

```diff
@@ -84,6 +84,14 @@ public sealed class CompletionItem : IComparable<CompletionItem>
         /// </summary>
         public CompletionItemRules Rules { get; }
 
+        /// <summary>
+        /// Returns true if this item's text edit requires complex resolution that
+        /// may impact performance. For example, an edit may be complex if it needs
+        /// to format or type check the resulting code, or make complex non-local
+        /// changes to other parts of the file.
+        /// </summary>
```

Comment:
annoying, but you tweak this to add a more affirmative statement that this work will be performed only for the commited item, and only just prior to commit.  It should say that this is thus a good place for putting in expensive work (like what you mentioned above) that is *not* necessary for what's put in the completion display, but is necessary to actually commit the item.